### PR TITLE
Some polish to testing

### PR DESCRIPTION
**Pull request by jvandort**
* Update TreeFormatter.appendMethod to also print parameter types. Before, the output was ambiguous if there were overloaded methods 
* Update TestSuite.getTargets to return a DOC instead of a polymorphic DOC. This is a generic type and we should not require all implementations to have a polymorphic collection of targets 
* Do not implement Named for TestSuiteTarget. Targets should not need to be uniquely named, as we have discussed in the past, different identities like java version can be used to ID a target 
* Add javadoc to many test event listener classes, clarifying what they do with test events 
* Rename DefaultTestReport to HtmlTestReport, as it was the only TestReporter implementation 
* Delete the TestReporter interface, as it was only used to inject mocked implementations during testing. Update Tests to not use mocks 
* Add comments to AbstractTestTask executeTests method, explaining the steps of test execution and which test event listeners are registered

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by jvandort** [comment]
@bot-gradle test this

**Comment by jvandort** [comment]
@bot-gradle test this

File: platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java [hunk 1]
```diff
@@ -1951,6 +1951,9 @@ private static Class<?> rawTypeParam(PropertyMetadata property, int paramNum) {
         if (argument instanceof Class) {
             return (Class<?>) argument;
         }
+        if (!(argument instanceof ParameterizedType)) {
+            throw new IllegalArgumentException("Argument is not an instance of ParameterizedType: " + argument);
```
Comment (big-guy) [inline platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java:1955]: You might consider `Cast.cast(...)`, but this is OK

File: platforms/jvm/testing-jvm/src/test/groovy/org/gradle/api/tasks/testing/TestTest.groovy [hunk 2]
```diff
@@ -106,14 +104,12 @@ class TestTest extends AbstractConventionTaskTest {
...
         configureTask()
-        final testReporter = Mock(TestReporter)
-        test.setTestReporter(testReporter)
 
         when:
         test.executeTests()
 
         then:
-        1 * testReporter.generateReport(_ as TestResultsProvider, reportDir)
+        reportDir.listFiles().collect { it.name }.contains("index.html")
```
Comment (big-guy) [inline platforms/jvm/testing-jvm/src/test/groovy/org/gradle/api/tasks/testing/TestTest.groovy:112]: I think you might be able to change `reportDir` to a `TestFile` and do:
```
reportDir.assertContainsDescendants("index.html")
```

**Review by octylFractal — APPROVED** [review]

File: platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java [hunk 3]
```diff
@@ -1951,7 +1952,7 @@ private static Class<?> rawTypeParam(PropertyMetadata property, int paramNum) {
         if (argument instanceof Class) {
             return (Class<?>) argument;
         }
-        return (Class<?>) ((ParameterizedType) argument).getRawType();
+        return (Class<?>) Cast.cast(ParameterizedType.class, argument).getRawType();
```
Comment (mlopatkin) [inline platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java:1955]: ❔ why do we need this change?

File: platforms/core-configuration/model-core/src/test/groovy/org/gradle/internal/instantiation/generator/AsmBackedClassGeneratorTest.java
Comment (mlopatkin) [inline platforms/core-configuration/model-core/src/test/groovy/org/gradle/internal/instantiation/generator/AsmBackedClassGeneratorTest.java:1]: 🎉  nice

**Review by mlopatkin — APPROVED** [review]
core-runtime lgtm

File: platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java (hunk 3 above)
Comment (jvandort) [reply platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AsmBackedClassGenerator.java:1955]: Late, but this gives a better error message when argument is not a ParameterizedType

**Pull request merged**